classes/Village.php

```diff
@@ -47,6 +95,33 @@ public function __construct($system, $village = '', array $village_row = []) {
             $this->coords = VillageManager::getLocation($this->system, $this->village_id);
             $this->relations = VillageManager::getRelationsForVillage($this->system, $this->village_id);
             $this->policy = new VillagePolicy($this->policy_id);
+            $this->buildings = VillageUpgradeManager::getBuildingsForVillage($this->system, $this->village_id);
+            $this->upgrades = VillageUpgradeManager::getUpgradesForVillage($this->system, $this->village_id);
+            $this->active_upgrade_effects = VillageUpgradeManager::initializeEffectsForVillage($this->system, $this->upgrades);
+            $this->construction_speed += $this->policy->construction_speed + $this->active_upgrade_effects[VillageUpgradeConfig::UPGRADE_EFFECT_CONSTRUCTION_SPEED];
+            $this->research_speed += $this->policy->research_speed + $this->active_upgrade_effects[VillageUpgradeConfig::UPGRADE_EFFECT_RESEARCH_SPEED];
+            if ($this->policy->construction_speed > 0) {
+                $result = $system->db->query("SELECT `village_id`, `construction_score` FROM `villages`");
```

Comment:
one other note, we're making two separate queries to `villages` here, we could consider combining this into one query with the field that `construction_speed` and `research_speed` calculations use and pre-loading it into this class